### PR TITLE
Add single node (per context) property cache and ZooKeeper storage

### LINKED ISSUE #1454: Create a single node (per context) for configuration in ZooKeeper
Configuration in ZooKeeper is currently stored in a tree of ZooKeeper nodes under some `conf/` directory. In order to reduce the number of watches, support atomic configuration changes in ZooKeeper, and generally make the configuration code simpler, we should serialize the configuration for each table, as well as the system-wide configuration, into a single node (each).

The serialization formation should be human-readable (JSON), and indicate a serialization format version (in case we change the serialized format later).

This is a sub-task of #1225.

This also relates to https://issues.apache.org/jira/browse/ACCUMULO-1568 and https://issues.apache.org/jira/browse/ACCUMULO-4052

### PR DESCRIPTION
**Pull request by EdColeman**
This PR partially resolves #1454.

Builds on PR #2224 that created VersionedProperties. This PR adds the
storgae and caching layer that will be used by the configuration
classes that manage the individual property nodes.

Additional IT testing will be added before final merge.

### CODE REVIEW

**Comment by EdColeman** [comment]
@keith-turner - are you still looking?  I've updated the code to use the micrometer metrics, but if I push, the merge could make it harder to follow comments to date and didn't want to just pull the rug out without warning.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  @Override
+  public @Nullable VersionedProperties load(PropCacheId propCacheId) {
+    try {
+      log.trace("load called for {}", propCacheId);
+      metrics.updateLoadCounter();
+      Stat stat = new Stat();
+      var vProps = propCodec.fromBytes(zrw.getData(propCacheId.getPath(), propStoreWatcher, stat));
+      if (stat.getVersion() != vProps.getDataVersion()) {
```
[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java:62]: This code is well structured in that it has dedicated code to go to/from Zookeeper <=> VersionProperties .  Code uses w/ the VersionProperties POJO which has a version, but the code using it does not have to care how or where that is persisted in ZK.  For correctness the code that persist to ZK has to care about what the version means relative to the stat object.  To me the code that persist the POJO to from ZK should be optimized for ZK as long as that does not leak outside to other code (like push ZK types out), which I don't think it would.

**Comment by keith-turner** [comment]
> @keith-turner - are you still looking? I've updated the code to use the micrometer metrics, but if I push, the merge could make it harder to follow comments to date and didn't want to just pull the rug out without warning.

I was still planning to look over it some more.  Go ahead and push the merge, I know functionally what I have looked at and have not.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+      throw new PropStoreException("read from prop store get() failed for: " + propCacheId, ex);
+    }
+  }
+
+  @Override
+  public boolean putAll(final PropCacheId propCacheId, final Map<String,String> props)
+      throws PropStoreException {
+
+    try {
+
```
[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:196]: I had thought about allowing retries on a conflict, but I could not reason out how to handle a conflict.  If both requests changed different things, then maybe it would be okay.

```
Thread-1 sets prop A to x
Thread-2 sets prop B to y

this seems "safe"
```

But what if it was a change?
```
prop A = z.
Thread-1 sets prop A = x
Thread-2 sets prop A = y
who wins?  If both return success, then what is the the value of prop A after both return?  Would you need to set and then call read to see if you "won"?
````

Rather than try to figure out logic that would be the least surprising, it seemed easiest to hard / quick fail and let the caller decide.  The calling function could make the decision to retry - it should have more context available to decide on what should happen in the case of a conflict.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: For the case of two thread setting the same thing concurrently we want to achieve the same result as them running serially in any order.  For the concurrent case if one succeeds and the other fails and retries, then we should get the same result as them running serially.

For example the following

```
Thread-1 reads the current props and stat from ZK
Thread-2 reads the current props and stat from ZK
Thread-1 modifies the props in memory to set A=x,B=z
Thread-2 modified the props in memory to set A=y,C=q
Thread-1 successfully writes the modified props to ZK
Thread-2 fails to write the props to ZK because the version does not match what it read
Thread-2 reads the current props and stat from ZK
Thread-2 modifies the props to set A=y,C=q
Thread-2 successfully writes the modified props to ZK
```

The above sequence of events should result in the props in ZK being `A=y,B=z,C=q` which is the same as what we would see if there was no concurrency in the following set of events. Neither one of these outcomes are surprising. 

```
Thread-1 reads the current props and stat from ZK
Thread-1 modifies the props in memory to set A=x,B=z
Thread-1 successfully writes the modified props to ZK
Thread-2 reads the current props and stat from ZK
Thread-2 modifies the props in memory to set A=y,C=q
Thread-2 successfully writes the modified props to ZK
```

[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: But what about:
```
Thread-1 reads the current props and stat from ZK
Thread-2 reads the current props and stat from ZK
Thread-1 modifies the props in memory to set A=x,B=z
Thread-2 modified the props in memory to set A=y,C=q
Thread-2 successfully writes the modified props to ZK
Thread-1 fails to write the props to ZK because the version does not match what it read
Thread-1 reads the current props and stat from ZK
Thread-1 modifies the props to set A=x,C=z
Thread-2 successfully writes the modified props to ZK
```
This would result in `A=x, B=z, C=q` not `A=y, B=z, C=q` in this case I would think that we would not want to retry without somehow evaluating if A=x or A=y is the desired value.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: > This would result in A=x, B=z, C=q not A=y, B=z, C=q in this case I would think that we would not want to retry without somehow evaluating if A=x or A=y is the desired value.

That is fine ` A=x, B=z, C=q` and `A=y, B=z, C=q` are both acceptable outcomes.  We want to achieve the same result as the operations running serially in any order.  Unacceptable outcomes of concurrency would be final zookeeper states of `A=x,B=z` or `A=y,C=q` because there is no serial ordering that would achieve those.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: For a more complex example if the following three set operations run concurrently.
  
 * `A=x,B=z`
 * `B=t,C=s`
 * `A=y,C=q`
 
Then the following is all of the ways these operations could run serially, the result of running them concurrently must result in one of these outcomes to be considered correct.

 * `A=x,B=z` then `B=t,C=s` then   `A=y,C=q` yielding `A=y,B=t,C=q`
 * `A=x,B=z` then `A=y,C=q` then `B=t,C=s`  yielding `A=y,B=t,C=s`
 * `A=y,C=q` then `A=x,B=z` then `B=t,C=s`  yielding `A=x,B=t,C=s`
 * `A=y,C=q` then `B=t,C=s` then `A=x,B=z`   yielding `A=x,B=z,C=s`
 * `B=t,C=s` then `A=y,C=q` then  `A=x,B=z`   yielding `A=x,B=z,C=q`
 * `B=t,C=s` then `A=x,B=z` then `A=y,C=q`    yielding `A=y,B=z,C=q`

[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: I disagree that any of them would be "correct". From an operation ordering viewpoint they are all consistent - but I argue that one of them represents the desired state.

As a contrived example, say two people use the shell and try to set the split threshold on a table, one picks 512M and the other uses 1G.  One of those values should be more correct than the other - which one would be left for them to figure out.  If both submitted the change such that they run concurrently, whatever the order that the operations run, one value will be the ultimate value of the property.  What I think you are suggesting is that both would receive success and neither would know which value was set without double checking.

For this level of the code - I'm proposing that one should succeed and the other should fail.  The one that receives the failure then has the option to either accept the change that got in first and "won", or they resubmit and change it to what they want.

The decision would be taken outside of this code - some operations may not care and elect to retry - others may choose throw an error - but it also leaves open other possibilities - maybe non-conflicting changes can be merged. That is, one sets X and the other sets Y, then setting both X and Y is acceptable. I didn't what to try and have that arbitration occur at this level.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: > What I think you are suggesting is that both would receive success and neither would know which value was set without double checking.

That situation can happen even w/o operations to set a property overlapping in time.  Consider the following.

 * via the shell user A makes an Accumulo API call to set the split threshold to 1G.  This call completes successfully w/o any other API set prop calls running concurrently.  The API call returns true and the shell stores this return value in a variable for user A's shell instance. 
 * via the shell user B makes an Accumulo API call to set the split threshold to 512M.  This call completes successfully w/o any other API set prop calls running concurrently.  The API call returns true and the shell stores this return value in a variable for user B's shell instance.
 * User A' shell prints success.
 * User B's shell prints success.

So even if the API code failed when concurrency was detected, it would not fail in this case because the concurrency happened outside the API call.  However the concurrency still happened and one user's set was overwritten by the other even though success was reported.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: The only way to deal the concurrency problem at the Accumulo API level would be to expose something like the version for getting and setting properties. The current APIs should probably retry.  If we have new APIs that expose this information, then those could fail if the version changed since reading the properties.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: Another way to deal with concurrency at the API level instead of exposing version information would be a new API that supports conditional set.  An API where a user could do something like set  `A=x,B=z` only if `A=r,B=y,C=m`.  This could be implemented on the backend using the version information w/o exposing the version information in the API.

[the remark this answers is not in the document]
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: This code is not public API and right now is strongly geared to configuration. I'd like to propose that we defer on this until the next layer PR is reviewed.  The next layer implements the configuration and if in that context enough calls would be simplified and benefit from just retrying I'll implement the retry here with the mutator.

[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:199]: >  The next layer implements the configuration and if in that context enough calls would be simplified and benefit from just retrying I'll implement the retry here with the mutator.

That's fine.  The discussion was good, I learned a few things from it.  Mainly I realized that our current API only supports unconditional set operations that do not care about any previous state.  Unconditional set operations should just retry when there is a failure because of concurrency because they have no concept of a required precondition. If we want to support conditional set operations, then those pre-conditions would have to be represented in the API in some way.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/CaffeineCache.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+  @Override
+  public @Nullable VersionedProperties get(PropCacheId propCacheId) {
+    log.trace("Called get() for {}", propCacheId);
+    try {
+      return cache.get(propCacheId);
+    } catch (Exception ex) {
+      log.info("Failed to get properties " + propCacheId, ex);
+      metrics.incrZkError();
+      return null;
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/CaffeineCache.java:84]: Why not propagate this exception?  Could do `throw new RuntimeException(e)`

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    return CacheLoader.super.asyncLoad(propCacheId, executor);
+  }
+
+  @Override
+  public CompletableFuture<? extends VersionedProperties> asyncReload(PropCacheId propCacheId,
+      VersionedProperties oldValue, Executor executor) throws Exception {
+    log.trace("asyncReload called for key: {}", propCacheId);
+    metrics.incrRefresh();
+
+    return CompletableFuture.supplyAsync(() -> loadIfDifferentVersion(propCacheId, oldValue));
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java:95]: `supplyAsync` may use a jvm wide shared executor service.  Seems like this can cause unrelated code in the JVM to interfere with each other, but I am not sure about this.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+    VersionedProperties vProps = new VersionedProperties(props);
+
+    try {
+
+      var path = propCacheId.getPath();
+      if (!zrw.putPersistentData(path, propCodec.toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL)) {
+        return false;
+      }
+
+      Stat stat = zrw.getStatus(path, propStoreWatcher);
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:166]: There may be a race condition here.  Something could run between the putPersistentData call and getStatus call that changes the stat to make it differ.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+    } catch (IOException | KeeperException | InterruptedException ex) {
+      throw new PropStoreException("Failed to serialize properties for " + propCacheId, ex);
+    }
+  }
+
+  @Override
+  public VersionedProperties get(final PropCacheId propCacheId) throws PropStoreException {
+    try {
+
+      checkZkConnection(); // if ZK not connected, block, do not just return a cached value.
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:185]: What happens if something calls and ZK is ok and the method returns, but after returning ZK becomes not ok before calling the following two methods?

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java:68]: > The only place where the ZooKeeper stat and the version need to be both known is on the ZooKeeper writes / reads - and I tried to isolate that to the ZooPropLoader.

ZooPropStore seems to also use the stat.

> Code that uses versioned props does not need to worry about the version. The version only matters on store / reads and then the ZooKeeper specifics are isolated (I hope) to ZooPropLoader - so I think I was trying to achieve the separation that you are asking for - or I'm missing what you are saying?

The separation is really nice, not advocating any change to that.  Just seems nicer to me if within the ZooPropLoader and ZooPropStore code that the Stat is the authoritative source of the persisted version not the version stored in the value.  So for write to ZK

```java
    zk.put(serializeEverythingButVersion(verProps), verProps.getVersion())
``` 

and read from ZK do the following.

```java
Stat stat = new stat;
byte data = zk.read(propPath, stat)
var verProp = newVersionedProps(data, stat)
```

If within these two classes ZK stat is the authority then do not need to do the following

 * check the version in the value against the stat in ZK after every read
 * make any assumptions about what the next version in ZK will be

**Comment by EdColeman** [comment]
@keith-turner I had similar discussion with @ctubbsii concerning the VersionedProperties being the source of truth for the version. With your concerns, it seems that I did not achieve the possible independence from the storage of the properties.  Certain things had to be ZooKeeper specific, but I was aiming that the storage at the byte array level could leverage something other than ZooKeeper if that was determined that it could provide advantages.

I'm now  thinking that this added complexity without a clear enough advantage at this time may be unwanted. I don't want to close this at this point.

I'm going to set this to draft and look at updating this PR to uses the stat version.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java (hunk 4 above)
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java:95]: ZooPropLoader is supplied to the Caffine cache and run in the context of the cache code.  There is a thread pool specified there and I think Caffeine will use that pool and not the general JVM pool.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/CaffeineCache.java (hunk 3 above)
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/CaffeineCache.java:84]: Certain parts of the Caffine code run asynchronously - if exceptions are thrown, they will only log the exception - returning null and then checking allowed those runtime exceptions to be handled outside of the caffeine cache.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java (hunk 4 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java:95]: > There is a thread pool specified there and I think Caffeine will use that pool and not the general JVM pool.

CompletableFuture.supplyAsync() is a java function that has no awareness of any caffeine thread pools.  Looking at the [javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#supplyAsync(java.util.function.Supplier)) it says.

> Returns a new CompletableFuture that is asynchronously completed by a task running in the ForkJoinPool.commonPool() with the value obtained by calling the given Supplier.

That ForkJoinPool.commonPool() is static JVM wide thread pool.

There is another version of [supply async that takes a thread pool](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#supplyAsync(java.util.function.Supplier,java.util.concurrent.Executor)).

Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java:95]: Great catch - the executor from the cache was being passed in to the function - I just neglected to use it here.

Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java]: Addressed in e8e60988

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java (hunk 5 above)
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:166]: I agree - but not sure what to do.  The creation of the node and then adding a watcher seems to require two ZooKeeper calls (or I'm missing something) There seems to be a few options:

1. Ignore the check. the getStatus call is only to set the watcher. It may not matter if it changed - the stored properties are only in ZooKeeper and not accessible until read and placed into the cache. That read would get whatever version is current.
2. Add a method in ZooReaderWriter that returns the Stat node from the ZooKeeper create call - and then check using that value.

Not sure if you have thoughts on which would be preferable.  This could be something that changes with using the ZooKeeper version instead of calculating it in VersionedProperties.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java (hunk 6 above)
Comment (EdColeman) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:185]: If the value is cached, it will be returned.  If it has not been previously seen, the ZooKeeper read from the caffeine loader will fail.

There is a watcher set that listens for connection events that will clear the cache on disconnection events so the time-window for returning a cached value should be narrow.  I don't think there is a way to avoid some window where a value can be read and returned and simultaneously a ZooKeeper disconnection event because the events are always asynchronous - so I tried to keep that window as small as possible, as well as have the disconnection event eventually recognized asynchronously.

**Comment by EdColeman** [comment]
@keith-turner I removed saving the version in the ZooKeeper blob and use the stat version as you suggested.

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java (hunk 6 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java:173]: What is the benefit of the check?

**Comment by ctubbsii** [comment]
This is replaced with #2569
